lib/everest.cpp

```diff
@@ -721,55 +722,26 @@ void Everest::external_mqtt_publish(const std::string& topic, const std::string&
-                                                    "set 'enable_external_mqtt' to 'true' in its manifest",
-                                                    this->config.printable_identifier(this->module_id))));
-    }
-
-    std::string external_topic = fmt::format("{}{}", this->mqtt_external_prefix, topic);
-
-    Handler external_handler = [handler, external_topic](const std::string&, json const& data) {
-        EVLOG_verbose << fmt::format("Incoming external mqtt data for topic '{}'...", external_topic);
-        if (!data.is_string()) {
-            EVLOG_AND_THROW(EverestInternalError("External mqtt result is not a string (that should never happen)"));
-        }
-        handler(data.get<std::string>());
-    };
-
-    std::shared_ptr<TypedHandler> token =
-        std::make_shared<TypedHandler>(HandlerType::ExternalMQTT, std::make_shared<Handler>(external_handler));
-    this->mqtt_abstraction.register_handler(external_topic, token, QOS::QOS0);
-    return [this, topic, token]() { this->mqtt_abstraction.unregister_handler(topic, token); };
+    auto external_topic = check_external_mqtt(topic);
+    return create_external_handler(
+        topic, external_topic, [handler, external_topic](const std::string&, json const& data) {
+            EVLOG_verbose << fmt::format("Incoming external mqtt data for topic '{}'...", external_topic);
+            if (!data.is_string()) {
+                EVLOG_AND_THROW(
+                    EverestInternalError("External mqtt result is not a string (that should never happen)"));
+            }
+            handler(data.get<std::string>());
+        });
 }
 
 UnsubscribeToken Everest::provide_external_mqtt_handler(const std::string& topic, const StringPairHandler& handler) {
     BOOST_LOG_FUNCTION();
-
-    // check if external mqtt is enabled
-    if (!this->module_manifest.contains("enable_external_mqtt") &&
-        this->module_manifest["enable_external_mqtt"] == false) {
-        EVLOG_AND_THROW(EverestApiError(fmt::format("Module {} tries to provide an external MQTT handler, but didn't "
-                                                    "set 'enable_external_mqtt' to 'true' in its manifest",
-                                                    this->config.printable_identifier(this->module_id))));
-    }
-
-    std::string external_topic = fmt::format("{}{}", this->mqtt_external_prefix, topic);
-
-    // must be json and not std::string
-    Handler external_handler = [handler](const std::string& topic, const json& data) {
+    auto external_topic = check_external_mqtt(topic);
+    return create_external_handler(topic, external_topic, [handler](const std::string& topic, const json& data) {
         EVLOG_verbose << fmt::format("Incoming external mqtt data for topic '{}'...", topic);
-        std::string data_s = (data.is_string()) ? std::string(data) : data.dump();
+        const std::string data_s = (data.is_string()) ? std::string(data) : data.dump();
```

Comment:
This is already much more readable :+1: .  Is there any reason why this function is allowed to have json data other than a string and the upper not?

Reply:
I can't comment on the original lambda - I'm not sure why there is a restriction to only support strings.
While testing the wildcard subscribe I found issues when the data portion was a JSON object and not just a string.
I can't see any reason to prevent the use of JSON with external mqtt, hence I provided support for it.